### DISCUSSION TITLE
What is our policy on 3rd party data collection for an integration as well as the method for asking for consent?

### DISCUSSION THREAD

**Discussion by raman325** [home-assistant/architecture#544] [General]
`zwave-js` recently introduced a capability to collect usage statistics [0][1] in a similar vein to the new analytics integration that was built for Home Assistant, and the question came up of how we should enable this for the `zwave_js` integration. We made some implicit statements and decisions around this, primarily that it must be opt-in, but it would be good to have an ADR that spells out our policies around this. Some things that should be a part of the ADR:
1. A statement about how 3rd party data collection *must* be opt-in (I'm assuming we all agree on this but we can debate this point if needed)
2. Guidance on how to present this consent screen to a user:
    1. Is adding consent to the options flow sufficient or do we need something else?
    2. I had an idea that we could start something like a `reauth` flow without failing the entry setup when the consent gets introduced to coax the user into making a decision about consent (but not block their usage of the integration). This is probably not the right approach since `reauth` is there for a specific use case, but should we consider supporting like this more explicitly?
    3. What should a consent option look like? Do we have a consent statement and a checkbox, do we use a drop down for yes/no, etc.

I'm just scratching the surface here so please add any additional questions that would be worth addressing and I will add them to the list.

[0]: https://zwave-js.github.io/node-zwave-js/#/data-collection/data-collection
[1]: https://zwave-js.github.io/node-zwave-js/#/data-collection/user-disclosure

BTW for additional context, here's what 2ii looks like to help visualize the problem:
Notification on startup: ![image](https://user-images.githubusercontent.com/7243222/114045829-54aa7a00-9856-11eb-8a34-1e5771eac3b0.png)
Integrations page: ![image](https://user-images.githubusercontent.com/7243222/114045864-5c6a1e80-9856-11eb-88cc-1c5143688cb1.png)
Click Reconfigure: 
![image](https://user-images.githubusercontent.com/7243222/114045916-65f38680-9856-11eb-825e-232a1e326490.png)

**Comment by MartinHjelmare** [comment]
We should discuss if want to allow this kind of option at all in the integration options and if so if there should be criteria to pass before acceptance.

**Comment by balloob** [comment]
The moment we add this to the UI and ask users, we are taking on responsibility what is gathered and what happens with this data because we call a library to gather and send that data.

We should not make it part of the config flow, nor trigger `reauth` to force users to make a decision. Instead, if we add it, it should be added to the Z-Wave JS panel. 

Also missing from your original description, the ZJS docs about this:
 - https://zwave-js.github.io/node-zwave-js/#/data-collection/data-collection
 - https://zwave-js.github.io/node-zwave-js/#/data-collection/user-disclosure

**Reply by raman325** [reply]
thanks, I added those to the description

**Reply by raman325** [reply]
btw, what's the issue with making it a part of the config flow? We would have to store the data on the config entry wouldn't we? Seems like unnecessary extra work to hide it from the standard and options config flow if it needs to be stored there

**Reply by balloob** [reply]
Config flows need to be limited to the bare minimum of information that we want to ask and get a user to use something. We should not ask extra things.

**Comment by cgarwood** [comment]
In the case of Z-Wave JS I wonder if it should be part of the addon config rather than integration config, especially since the data goes to a third party. Though that could still require something in the config flow since we offer to setup the addon automatically?

Doesn't help integrations that may want to collect analytics without an addon (thinking something like ZHA) though.

**Reply by frenck** [reply]
I think that makes sense to have in the add-on. For example, if a user is using the Z-Wave2MQTT Docker container or add-on, they would also already have been asked for this?

It seems like in those cases, the UI of Z-Wave2MQTT controls / keeps the state of the opt-in/opt-out (and also stores the preference)

Maybe that is a design flaw, as the state opt-in/out should be held by Z-WaveJS itself, not the application layer.

**Reply by balloob** [reply]
Actually it's by design. From [the docs](https://zwave-js.github.io/node-zwave-js/#/data-collection/data-collection):

> This information is only collected if the application developer has opted in into this functionality. 

The driver itself is not storing configuration, it's passed in.

**Reply by frenck** [reply]
The problem is that the current design would allow for different settings in multiple applications that use the same Z-Wave JS server.. which is used and right...?
